java/code/src/com/suse/manager/clusters/ClusterManager.java

```diff
@@ -0,0 +1,609 @@
+     * @throws TaskomaticApiException if the system group refresh could not be schedules
+     */
+    public Cluster addCluster(String name, String label, String description, long managementNodeId,
+                              String provider, Map<String, Object> managementSettings, User user)
+            throws IOException, TaskomaticApiException {
+        Server managementNode = ServerFactory.lookupById(managementNodeId);
+        if (managementNode == null) {
+            throw new RuntimeException("Server with id=" + managementNodeId + " not found");
+        }
+        if (managementNode.asMinionServer().isEmpty()) {
+            throw new RuntimeException("Server with id=" + managementNodeId + " is not a minion");
+        }
+
+        // find out settings formula from provider metadata
+        String settingsFormula = FormulaFactory.getClusterProviderMetadata(provider,
+                "formulas:settings:name", String.class)
+                .orElseThrow(() -> new RuntimeException(
+                        "Missing key formulas:settings:name from metadata.yml of cluster provider " +
+                        provider));
+
+        // create corresponding group
+        ManagedServerGroup group = serverGroupManager.create(user, getGroupName(name),
+                getGroupDescription(label));
+
+        // create cluster
+        Cluster cluster = new Cluster();
+        cluster.setName(name);
+        cluster.setLabel(label);
+        cluster.setDescription(description);
+        cluster.setProvider(provider);
+        cluster.setManagementNode(managementNode.asMinionServer().get());
+        cluster.setGroup(group);
+        cluster.setOrg(user.getOrg());
+        getSession().save(cluster);
+
+        // enable settings formula
+        List<String> formulas = new ArrayList<>();
+        formulas.addAll(FormulaFactory.getFormulasByGroupId(group.getId()));
+        formulas.add(settingsFormula);
+        FormulaFactory.saveGroupFormulas(group.getId(), formulas, user.getOrg());
+
+        // save settings data
+        Map<String, Object> settingsInNamespace = adjustNamespace(label, "settings", managementSettings);
+        FormulaFactory.saveGroupFormulaData(settingsInNamespace, group.getId(), user.getOrg(), settingsFormula);
+
+        // add management node to group
+        serverGroupManager.addServers(group, Arrays.asList(managementNode), user);
+
+        // schedule refreshing group nodes
+        ClusterActionCommand<ClusterGroupRefreshNodesAction> clusterActionCommand =
+                new ClusterActionCommand(Optional.of(user), user.getOrg(),
+                        new Date(),
+                        null,
+                        ActionFactory.TYPE_CLUSTER_GROUP_REFRESH_NODES,
+                        managementNode.asMinionServer().get(),
+                        cluster,
+                        cluster.getGroup().getName(),
+                        null);
+        clusterActionCommand.store();
```

Comment:
If possible, the whole thing should be done in a single DB transaction

Reply:
Are you referring to refresh group nodes ? It must query the cluster so it can take some time, that's why it's done async.

Reply:
No, I mean:
 - Create the server group
 - Create the cluster
 - Create and store the action entry
These should be all-or-nothing